tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,411 @@
+        "The resonance finder is initialized with an even grid of frequencies between "
+        "fmin and fmax. The resonance finder then iteratively optimizes and prunes these "
+        "frequencies. Note that frequencies outside this window may be returned. "
+        "A narrow frequency window that only contains a few resonances may give enhanced "
+        "accuracy compared to a broad frequency window with many resonances.",
+        units=HERTZ,
+    )
+
+    num_freqs_bounds: Tuple[PositiveInt, PositiveInt] = Field(
+        (NUM_FREQ_BOUND_MIN, NUM_FREQ_BOUND_MAX),
+        title="Bounds ``[nmin, nmax]`` on the initial number of frequencies.",
+        description="Bounds ``[nmin, nmax]`` on the initial number of frequencies. "
+        "The resonance finder is initialized with ``max(nmin, min(nmax, nguess))`` frequencies, "
+        "where ``nguess = signal_length*time_step*(fmax - fmin)/2``. The resonance finder then "
+        "iteratively optimizes and prunes these frequencies. Note that the number of frequencies "
+        "returned may be less than ``nmin`` but will not exceed ``nmax``. "
+        "Making these bounds larger can find more resonances at the expense of run-time "
+        "and matrix conditioning.",
+    )
+
+    rcond: NonNegativeFloat = Field(
+        1e-4,
+        title="Cutoff for eigenvalues",
+        description="Cutoff for eigenvalues, relative to the largest eigenvalue. "
+        "The algorithm requires solving a generalized eigenvalue problem of the form "
+        "Ax = lambda B x. If B has small eigenvalues, this is poorly conditioned, "
+        "so we eliminate eigenvalues of B less than ``rcond`` times the largest eigenvalue. "
+        "Making this closer to one will typically return fewer resonances, while making it "
+        "closer to zero will typically return more resonances.",
+    )
+
+    @validator("freq_window", always=True)
+    def _check_freq_window(cls, val):
+        """Validate ``freq_window``"""
+        if val[1] < val[0]:
+            raise ValidationError(
+                "The value of ``freq_window[1]`` is less than the value of ``freq_window[0]``."
+            )
+        return val
+
+    @validator("num_freqs_bounds", always=True)
+    def _check_num_freqs_bounds(cls, val):
+        """Validate ``num_freqs_bounds``"""
+        if val[1] < val[0]:
+            raise ValidationError(
+                "The value of ``num_freqs_bounds[1]`` is less than "
+                "the value of ``num_freqs_bounds[0]``."
+            )
+        return val
+
+    def run(self, signals: Union[FieldTimeData, Tuple[FieldTimeData, ...]]) -> xr.Dataset:
+        """Finds resonances in a 'FieldTimeData' or a Tuple of such.
+        Time time coordinates must be uniformly spaced, and the spacing must be the same
+        between all supplied 'FieldTimeData'. The resonance finder runs on the sum of the
+        'Ex', 'Ey', and 'Ez' fields of all supplied 'FieldTimeData', if they exist, or else
```

Comment:
I don't know if you ever saw [this page](https://www.notion.so/flexcompute/Python-Style-Guide-afac5dbbf949463e89c18fe39b7453c9), but have a look there. It's a bit confusing at first and may not be thoroughly implemented throughout the code, but we're trying to stick to this for new code. In particular, anywhere in the docstrings where you have something like `'Ex'`, it should be 
```
``Ex``
```
instead (for good reasons).